### ISSUE TITLE
CDEF of Type Special Data Source not working as expected

### ISSUE THREAD

**Issue by j66h** [Cacti/cacti#2971]
**Environment**
OS: Windows Server 2016 standard
Cacti: 1.2.6  (updated to latest 1.2.x branch as of 09/23/2019)
Spine: 1.2.6
Cygwin: 3.0.7
Net-SNMP: 5.7.3
PHP: 7.3.9
MySQL: 8.0.17

Single server. Just "Main Poller".

**Describe the bug**
Graphs, which use CDEF of type "Special Data Source" do not render. 

Tested with several of the special types.
- All_DATA_SOURCES_DUP
- All_DATA_SOURCES_NODUP
- SIMILAR_DATA_SOURCES_DUPS
- SIMILAR_DATA_SOURCES_NODUPS
- ...

All these types lead to an empty CDEF shown in Graph Debug mode:
CDEF:cdefa= \

Graph uses two data sources. CDEF should be something like CDEF:cdefa="a,b,+" \

**Screenshots**
![2019-09-23 13_49_13-Console -_ Graph Templates -_ (Edit) -_ Graph Template Items](https://user-images.githubusercontent.com/54892435/65423431-3cf0d980-de09-11e9-91d2-93e9547c8810.png)

Built in CDEF "Total All Data Sources" uses special data source All_DATA_SOURCES_NODUP

**Additional context**
Call to rrdtool.exe from Graph Debug mode
`c:/rrdtool/rrdtool.exe graph - \
--imgformat=PNG \
--start="-86400" \
--end="-300" \
--pango-markup  \
--title="Hostname - Cisco Memory Usage" \
--slope-mode \
--base=1024 \
--height=120 \
--width=500 \
--tabwidth "30" \
--alt-autoscale-max \
--lower-limit="0" \
COMMENT:"From 22.09.2019 12\:47\:38 To 23.09.2019 12\:42\:38\c" \
COMMENT:"  \n" \
--color BACK#F3F3F3 \
--color CANVAS#FDFDFD \
--color SHADEA#CBCBCB \
--color SHADEB#999999 \
--color FONT#000000 \
--color AXIS#2C4D43 \
--color ARROW#2C4D43 \
--color FRAME#2C4D43 \
--border 1 --font TITLE:11:"Arial" \
--font AXIS:8:"Arial" \
--font LEGEND:8:"Courier" \
--font UNIT:8:"Arial" \
--font WATERMARK:6:"Arial" \
--slope-mode \
--watermark "Watermark«" \
DEF:a="D\:/inetpub/wwwroot/cacti/rra/hostname_mem_used_12975.rrd":"mem_used":MAX \
DEF:b="D\:/inetpub/wwwroot/cacti/rra/hostname_mem_free_12974.rrd":"mem_free":MAX \
CDEF:cdefa= \
AREA:cdefa#750F7DFF:"Total Available Memory\:\n"  \
GPRINT:cdefa:LAST:"Available Memory\:%8.2lf %s\n"  \
LINE2:a#00FF00FF:"Used Memory\:           \n"  \
GPRINT:a:LAST:"Current\:%8.2lf %s"  \
GPRINT:a:MAX:"Max\:%8.2lf %s"  \
GPRINT:a:AVERAGE:"Average\:%8.2lf %s\n"  \
LINE2:b#F5F800FF:"Free Memory\:           \n"  \
GPRINT:b:LAST:"Current\:%8.2lf %s"  \
GPRINT:b:MAX:"Max\:%8.2lf %s"  \
GPRINT:b:AVERAGE:"Average\:%8.2lf %s\n" `

**Comment by cigamit** [comment]
Is there anything in your Cacti log?

**Comment by cigamit** [comment]
Also, which template is this?

**Comment by j66h** [comment]
This one is "Cisco Memory Template" which might be the original one.
But this cacti installation lives since 15 years now and was operated by several people. So I cannot be sure on this...

Nothing in Cacti log. Do you have a suggestion for specific debugs to turn on?

I definded a CDEF for "a+b" myself. Graph immediately is rendered. So RRD contains valid data...

Error graph while set for "Total All Data Sources"
![Error](https://user-images.githubusercontent.com/54892435/65424923-0ddc6700-de0d-11e9-8648-2232b3e7ff28.png)
But Graph Debug shows "RRDtool Says: OK" while this picture is shown.

Graph while CDEF:cdefa="a,b,+" is in use:
![Success](https://user-images.githubusercontent.com/54892435/65424959-2c426280-de0d-11e9-9a92-4272700edca9.png)

**Comment by cigamit** [comment]
It's quite odd.  Have you attempted to 'sync' the graphs to the template?  I have found a bug where if you attempt to total on a CF that is not on a graph, it only get's one.  So, for example, say you have AVG and MAX on the graph, and then attempt to total on all data sources, but use MIN, then you only get one, but I can not reproduce your issue yet.  I've got some hunches.  This code is pretty old stuff.  Might be cross RRDfile the issue.  Going to test with another Graph Template that pull in Data Sources from three separate RRDfiles now.

**Comment by cigamit** [comment]
Second observation is that across three RRDfiles, that it's using MAX even though I'm selecting average.  Seems a bit buggy.  Though I can not re-produce your case yet.  Looks like this will be the goal for the next week (in other words before the 1.2.7 release) as verification of all these is going to take time.

**Comment by j66h** [comment]
Recreated some graphs from scratch while troubleshooting. (deleted graph from device including data source and afterwards recreated graph for same device).

Created a new template and configured it the same like the old one. Started just with mem_free and mem_used. As soon as I added "Total All Data Sources" Error graph was shown. In that new template I started playing arround with CDEF "a+b". After that worked, I started playing with the old template. My own "a+b" worked there too. Changed to different Special Types. Only types like "Graph: lower limit" worked. All types which sum up something didn't work.

Will try another template which I defined short ago.

**Comment by j66h** [comment]
I think, this is no high prio issue, since one manually can define a sum of all data sources like described above. So take your time :-)

Tried to understand the code, where it sets the CDEFS, but only was partially successful :-(

**Comment by j66h** [comment]
Found something.
Seems like problems  arise, if DEF has "MAX" at its end.

Then CDEF is empty.
`DEF:a="D\:/inetpub/wwwroot/cacti/rra/hostname_mem_used_12975.rrd":"mem_used":MAX \
DEF:b="D\:/inetpub/wwwroot/cacti/rra/hostname_mem_free_12974.rrd":"mem_free":MAX \
CDEF:cdefa= \
`

If I have AVERAGE at the end, everything looks ok.
`DEF:a="D\:/inetpub/wwwroot/cacti/rra/hldehkr36_mem_free_13003.rrd":"mem_free":AVERAGE \
DEF:b="D\:/inetpub/wwwroot/cacti/rra/hldehkr36_mem_used_13004.rrd":"mem_used":AVERAGE \
CDEF:cdefc="TIME,1569246272,GT,a,a,UN,0,a,IF,IF,TIME,1569246272,GT,b,b,UN,0,b,IF,IF,+" \`

Maybe I'm blind today, but I do not see, where I could set this...

**Comment by cigamit** [comment]
It's more than that.  These 'special functions' have not been audited in some time.  So, they are going to need a once over.  Like you said, maybe not a rush.  But I did, in some testing, find some oddities.

**Comment by cigamit** [comment]
So, I've created a fix for another problem that appears to resolve this problem.  But the interesting thing is that the total all data sources wants to bias the Consolidation Function (CF) to the first one it finds in the Graph.  So, if you have a mixture of MAX and AVERAGE DEF's, it does not total correctly.

So, this seems to be a design issue.  But the bug you were experiencing has been fixed, at least for the SAME CF.  I will open a separate ticket for the issue with differing CF's on the DEF statement, and mark duplicate with the other ticket shortly.  Need to update the CHANGELOG though.
